OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1413 @@
+                    originalUseSites: mutable.Map[Int, List[Int]]
+                  ): Boolean = {
+        params.foreach { p ⇒
+            if (p.asVar.value.toString.startsWith(intentStart.head) || p.asVar.value.toString.startsWith(intentStart(1))) {
+                //the intent to its source
+                val defIndex = p.asVar.definedBy.head
+                if (defIndex < 0) {
+                    return false //intent is not reachable
+                }
+                if (originalUseSites.exists(k ⇒ k._2.contains(defIndex))) {
+                    return true
+                }
+                val stm = tacCode.stmts(defIndex)
+                if (stm.isAssignment) {
+                    if (stm.asAssignment.expr.isStaticFunctionCall || stm.asAssignment.expr.isVirtualFunctionCall) {
+                        return findOrigin(stm.asAssignment.expr.asFunctionCall.params, tacCode, originalUseSites)
+                    }
+                }
+            }
+        }
+        false //intent is not reachable
+    }
+
+    /**
+     * searches the parameters of functionCall to find the origin of the intent and return the class and method
+     * that returns the searched intent
+     *
+     */
+    def searchParamsStatic(
+                            functionCall: FunctionCall[DUVar[ValueInformation]],
+                            tacCode: AITACode[TACMethodParameter, ValueInformation]
+                          ): Option[(ObjectType, String)] = {
+        functionCall.params.foreach { p ⇒
+            if (p.asVar.value.toString.startsWith(intentStart.head) || p.asVar.value.toString.startsWith(intentStart(1))) {
+                val defIndex = p.asVar.definedBy.head
+                if (defIndex < 0) {
+                    return None
+                }
+                val defSiteStm = tacCode.stmts(defIndex)
+                if (defSiteStm.isAssignment) {
+                    val call = defSiteStm.asAssignment.expr.asFunctionCall
+                    if (!intentMethods.contains(call.name)) {
+                        return Some[(ObjectType, String)](call.declaringClass.asObjectType -> call.name)
+                    }
+                    return searchParamsStatic(call, tacCode)
+                }
+            }
+        }
+        Some[(ObjectType, String)](functionCall.declaringClass.asObjectType -> functionCall.name)
+    }
+
+    /**
+     * searches the parameters of functionCall to find the origin of the intent and return the class and method
+     * that returns the searched intent
+     *
+     */
+    def searchParamsVirtual(
+                             functionCall: MethodCall[DUVar[ValueInformation]],
+                             tacCode: AITACode[TACMethodParameter, ValueInformation]
+                           ): Option[(ObjectType, String)] = {
+        functionCall.params.foreach { p ⇒
+            if (p.asVar.value.toString.startsWith(intentStart.head) || p.asVar.value.toString.startsWith(intentStart(1))) {
```

Comment:
You should probably extract p.asVar.value.toString to a local variable to shorten this.